libbpf-tools/biolatency.bpf.c

```diff
@@ -52,18 +53,34 @@ int trace_rq_start(struct request *rq)
 }
 
 SEC("tp_btf/block_rq_insert")
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 11, 1)
```

Comment:
this can't be done at the compilation time, because Linux version on which you compile BPF code has nothing to do with the kernel version that it will be run on. There are two possible approaches here.

First, is to do what you did here (two sets of BPF programs), but without any compile-time `#if`/`#else`. Then from user-space side detect which set of BPF programs needs to be used based on actual running kernel support. And then bpf_program__set_autoload(false) for the BPF programs that won't work on the kernel. This is most flexible approach, but requires quite a bit more code.

Alternatively, libbpf supports .kconfig externs, so one can do kernel version check in runtime in BPF code. So the solution would be like this:

```
#define KERNEL_VERSION(a, b, c) (((a) << 16) + ((b) << 8) + (c))

extern int LINUX_KERNEL_VERSION __kconfig;

SEC("tp_btf/block_rq_insert")
int block_rq_insert(u64 *ctx)
{
    /* comment here pointing to the commit that changed tracepoint argument list */
    struct request *rq = LINUX_KERNEL_VERSION >= KERNEL_VERSION(5, 11, 1) ? ctx[1] : ctx[0];

    /* the rest is untouched */
}
```

libbpf will substitute actual runtime kernel version into LINUX_KERNEL_VERSION extern.

Making such decisions based on kernel version is a bit bad, because there are kernels that get a lot of patches backported, but kernels themselves stay on old kernel version. But given the minimal amount of code, it might be a good start (and a first demo of .kconfig externs in libbpf-tools ;) ).